packages/db/tests/models/test-granule-model.js

```diff
@@ -244,11 +475,314 @@ test('GranulePgModel.upsert() will allow a running status to replace a completed
+    knex,
+    granule
+  );
+
+  t.like(
+    granuleRecord,
+    {
+      ...granule,
+      cumulus_id: granuleCumulusId,
+    }
+  );
+  t.deepEqual(
+    await granuleExecutionHistoryPgModel.search(
+      knex,
+      { granule_cumulus_id: granuleCumulusId }
+    ),
+    [executionCumulusId, secondExecutionCumulusId].map((executionId) => ({
+      granule_cumulus_id: Number.parseInt(granuleCumulusId, 10),
+      execution_cumulus_id: executionId,
+    }))
+  );
+});
+
+test('GranulePgModel.upsertWithExecutionHistory() works with transaction', async (t) => {
+  const {
+    knex,
+    granulePgModel,
+    collectionCumulusId,
+    executionCumulusId,
+  } = t.context;
+
+  const granule = fakeGranuleRecordFactory({
+    collection_cumulus_id: collectionCumulusId,
+    status: 'running',
+  });
+
+  const [granuleCumulusId] = await knex.transaction(
+    (trx) =>
+      granulePgModel.upsertWithExecutionHistory(
+        trx,
+        granule,
+        executionCumulusId
+      )
+  );
+
+  const [granuleWithHistory] = await granulePgModel.getWithExecutionHistory(
+    knex,
+    granule
+  );
+
+  t.like(
+    granuleWithHistory,
+    {
+      ...granule,
+      cumulus_id: granuleCumulusId,
+      execution_cumulus_ids: [executionCumulusId],
+    }
+  );
+});
+
+test('GranulePgModel.upsertWithExecutionHistory() does not write anything if execution history upsert fails', async (t) => {
+  const {
+    knex,
+    granulePgModel,
+    collectionCumulusId,
+    executionCumulusId,
+    granuleExecutionHistoryPgModel,
+  } = t.context;
+
+  const granule = fakeGranuleRecordFactory({
+    collection_cumulus_id: collectionCumulusId,
+    status: 'running',
+  });
+
+  const fakeGranuleExecutionHistoryPgModel = {
+    upsert: async () => {
+      throw new Error('error');
+    },
+  };
+
+  await t.throwsAsync(
+    knex.transaction(
+      (trx) =>
+        granulePgModel.upsertWithExecutionHistory(
+          trx,
+          granule,
+          executionCumulusId,
+          fakeGranuleExecutionHistoryPgModel
+        )
+    )
+  );
+
+  t.false(
+    await granulePgModel.exists(
+      knex,
+      {
+        granule_id: granule.granule_id,
+        collection_cumulus_id: collectionCumulusId,
+      }
+    )
+  );
+  t.false(
+    await granuleExecutionHistoryPgModel.exists(
+      knex,
+      {
+        execution_cumulus_id: executionCumulusId,
+      }
+    )
+  );
+});
+
+test('GranulePgModel.upsertWithExecutionHistory() will allow a running status to replace a completed status for different execution', async (t) => {
+  const {
+    knex,
+    granulePgModel,
+    executionPgModel,
+    granuleExecutionHistoryPgModel,
+    collectionCumulusId,
+    executionCumulusId,
+  } = t.context;
+
+  const granule = fakeGranuleRecordFactory({
+    collection_cumulus_id: collectionCumulusId,
+    status: 'completed',
+  });
+
+  const [granuleCumulusId] = await granulePgModel.upsertWithExecutionHistory(
```

Comment:
Nit: By using the logic of the method under test to setup the test for the method being tested it's possible to setup a situation where the method under test undermines the logic you're trying to test. If the other units adequately cover it, you're fine, thus the nit, but consider using the create or upsert model methods directly for explicitness.